bot/engine/src/main/kotlin/engine/Bot.kt

```diff
@@ -108,26 +111,28 @@ internal class Bot(botDefinitionBase: BotDefinition, val configuration: BotAppli
 
         if (!userTimeline.userState.botDisabled) {
             dialog.state.currentIntent?.let { intent ->
-                connector.sendIntent(intent, action.applicationId, connectorData)
+                connector.sendIntent(intent, action.connectorId, connectorData)
             }
             connector.startTypingInAnswerTo(action, connectorData)
             val story = getStory(userTimeline, dialog, action)
             val bus = TockBotBus(connector, userTimeline, dialog, action, connectorData, botDefinition)
-
-            if (bus.isFeatureEnabled(DefaultFeatureType.DISABLE_BOT)) {
-                logger.info { "bot is disabled for the application" }
-                bus.end("Bot is disabled")
-                return
-            }
-
-            try {
-                currentBus.set(bus)
-                story.handle(bus)
-                if (shouldRespondBeforeDisabling) {
-                    userTimeline.userState.botDisabled = true
+            val asyncBus = AsyncBotBus(bus)
+
+            withContext(AsyncBotBus.Ref(asyncBus) + currentBus.asContextElement(bus)) {
+                val closeMessageQueue = bus.deferMessageSending(this)
+
+                if (asyncBus.isFeatureEnabled(DefaultFeatureType.DISABLE_BOT)) {
+                    logger.info { "bot is disabled for the application" }
+                    asyncBus.end("Bot is disabled")
+                } else {
+                    story.handle(asyncBus)
+                    if (shouldRespondBeforeDisabling) {
+                        userTimeline.userState.botDisabled = true
+                    }
                 }
-            } finally {
-                currentBus.remove()
+
+                // Ensure we do not have a lingering message sending job
+                closeMessageQueue()
```

Comment:
May be a a try finally is better here

Reply:
I was somewhat confused, there is no special interaction between cancellation and channels, but it should nonetheless be fine: https://stackoverflow.com/questions/43889066/what-closing-a-kotlinx-coroutines-channel-does